src/extra/sprite.cpp

```diff
@@ -0,0 +1,406 @@
+       {CoreInfo::AnyTableType, CoreInfo::AnyVarTableType}}};
+
+  // params
+  ParamVar _atlas{};
+
+public:
+  struct Region {
+    uint32_t width;
+    uint32_t height;
+    float uvs[4];
+  };
+
+  uint32_t width;
+  uint32_t height;
+  bool premultiply;
+  std::vector<Region> regions;
+};
+
+struct Draw {
+  static CBTypesInfo inputTypes() { return CoreInfo::Float4x4Type; }
+
+  static CBTypesInfo outputTypes() { return CoreInfo::Float4x4Type; }
+
+  static CBParametersInfo parameters() { return _params; }
+
+  void setParam(int index, const CBVar &value) {
+    switch (index) {
+    case 0:
+      _sheet = value;
+      break;
+    case 1:
+      _texture = value;
+      break;
+    case 2:
+      _index = value;
+      break;
+    case 3:
+      _shader = value;
+      break;
+    case 4:
+      _blend = value;
+      break;
+
+    default:
+      throw CBException("Parameter out of range.");
+    }
+  }
+
+  CBVar getParam(int index) {
+    switch (index) {
+    case 0:
+      return _sheet;
+    case 1:
+      return _texture;
+    case 2:
+      return _index;
+    case 3:
+      return _shader;
+    case 4:
+      return _blend;
+    }
+    throw CBException("Parameter out of range.");
+  }
+
+  void cleanup() {
+    _sheet.cleanup();
+    _texture.cleanup();
+    _index.cleanup();
+    _shader.cleanup();
+  }
+
+  void warmup(CBContext *context) {
+    _sheet.warmup(context);
+    _texture.warmup(context);
+    _index.warmup(context);
+    _shader.warmup(context);
+
+    _bgfxContext = referenceVariable(context, "GFX.Context");
+  }
+
+  CBVar activate(CBContext *context, const CBVar &input) {
+    if (input.payload.seqValue.len != 4) {
+      throw ActivationError("Invalid Matrix4x4 input, should Float4 x 4.");
+    }
+
+    auto *sheet = reinterpret_cast<Sheet *>(_sheet.get().payload.objectValue);
+
+    auto *ctx =
+        reinterpret_cast<BGFX::Context *>(_bgfxContext->payload.objectValue);
+    const auto &currentView = ctx->currentView();
+
+    // step: transform
+    bgfx::Transform t;
+    // using allocTransform to avoid an extra copy
+    auto idx = bgfx::allocTransform(&t, 1);
+    memcpy(&t.data[0], &input.payload.seqValue.elements[0].payload.float4Value,
+           sizeof(float) * 4);
+    memcpy(&t.data[4], &input.payload.seqValue.elements[1].payload.float4Value,
+           sizeof(float) * 4);
+    memcpy(&t.data[8], &input.payload.seqValue.elements[2].payload.float4Value,
+           sizeof(float) * 4);
+    memcpy(&t.data[12], &input.payload.seqValue.elements[3].payload.float4Value,
+           sizeof(float) * 4);
+    bgfx::setTransform(idx, 1);
+
+    // step: create quad
+    auto index = _index.get().payload.intValue;
+    const auto &region = sheet->regions[index % sheet->regions.size()];
+    bgfx::TransientVertexBuffer vb;
+    // FIXME: bound/size doesn't work if the sprite is rotated
+    createQuad(&vb, region.width / (float)currentView.width,
+               region.height / (float)currentView.height);

```

Comment:
This division is not correct if a rotation is applied to the sprite.

For example in `src\tests\sprite_test.edn`, if we change the transform to be
```clojure
90.0 (Math.DegreesToRadians) (Math.AxisAngleZ) (Math.Rotation) = .r
```

I believe we must used the projection matrix of the current view or something similar, but I'm not sure of the math involved.

Reply:
Addressed in sprite_camera branch